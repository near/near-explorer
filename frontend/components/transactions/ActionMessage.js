import { formatNEAR } from "../utils/Balance";

const transactionMessageRenderers = {
  CreateAccount: ({ transaction: { signerId } }) =>
    `New Account Created: @${signerId}`,
  DeleteAccount: ({ transaction: { signerId } }) =>
    `Account Deleted: @${signerId}`,
  DeployContract: ({ transaction: { signerId } }) =>
    `Contract Deployed: ${signerId}`,
  FunctionCall: ({ transaction: { receiverId } }) =>
    `Call: Called method in contract "${receiverId}"`,
  Transfer: ({ transaction: { receiverId }, actionArgs: { deposit } }) =>
    `Transfered ${formatNEAR(deposit)} Ⓝ to @${receiverId}`,
  Stake: ({ actionArgs: { stake, public_key } }) =>
    `Staked: ${formatNEAR(stake)} Ⓝ ${public_key.substring(0, 15)}...`,
  AddKey: ({ transaction: { receiverId }, actionArgs }) => {
    return actionArgs.access_key
      ? `Access key for contract: "${receiverId}"`
      : `New Key Created: ${actionArgs.publicKey}`;
  },
  DeleteKey: ({ actionArgs: { public_key } }) => `Key Deleted: ${public_key}`
};

export default props => {
  const MessageRenderer = transactionMessageRenderers[props.actionKind];
  if (MessageRenderer === undefined) {
    return `${props.actionKind}: ${JSON.stringify(props.actionArgs)}`;
  }
  return <MessageRenderer {...props} />;
};
