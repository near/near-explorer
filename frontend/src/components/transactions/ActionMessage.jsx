import { formatNEAR } from "../utils/Balance";

const transactionMessageRenderers = {
  CreateAccount: ({ transaction: { receiverId } }) =>
    `New Account Created: @${receiverId}`,
  DeleteAccount: ({ transaction: { receiverId } }) =>
    `Account Deleted: @${receiverId}`,
  DeployContract: ({ transaction: { receiverId } }) =>
    `Contract Deployed: ${receiverId}`,
  FunctionCall: ({ transaction: { receiverId } }) =>
    `Call: Called method in contract "${receiverId}"`,
  Transfer: ({ transaction: { receiverId }, actionArgs: { deposit } }) =>
    `Transfered ${formatNEAR(deposit)} Ⓝ to @${receiverId}`,
  Stake: ({ actionArgs: { stake, public_key } }) =>
    `Staked: ${formatNEAR(stake)} Ⓝ ${public_key.substring(0, 15)}...`,
  AddKey: ({ transaction: { receiverId }, actionArgs }) => {
    return typeof actionArgs.access_key.permission === "object"
      ? `Access key for contract: "${
          actionArgs.access_key.permission.FunctionCall.receiver_id
        }"`
      : `New Key Added for @${receiverId}: ${actionArgs.public_key.substring(
          0,
          15
        )}...`;
  },
  DeleteKey: ({ actionArgs: { public_key } }) =>
    `Key Deleted: ${public_key.substring(0, 15)}...`
};

export default props => {
  const MessageRenderer = transactionMessageRenderers[props.actionKind];
  if (MessageRenderer === undefined) {
    return `${props.actionKind}: ${JSON.stringify(props.actionArgs)}`;
  }
  return <MessageRenderer {...props} />;
};
