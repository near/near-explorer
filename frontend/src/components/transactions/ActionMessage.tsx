import { hexy } from "hexy";

import AccountLink from "../utils/AccountLink";
import { formatNEAR } from "../utils/Balance";
import CodePreview from "../utils/CodePreview";

import * as T from "../../libraries/explorer-wamp/transactions";

export interface Props<A> {
  actionKind: keyof TransactionMessageRenderers;
  actionArgs: A;
  transaction: T.Transaction;
  showDetails?: boolean;
}

type AnyAction =
  | T.CreateAccount
  | T.DeleteAccount
  | T.DeployContract
  | T.FunctionCall
  | T.Transfer
  | T.Stake
  | T.AddKey
  | T.DeleteKey;

interface TransactionMessageRenderers {
  CreateAccount: React.FC<Props<T.CreateAccount>>;
  DeleteAccount: React.FC<Props<T.DeleteAccount>>;
  DeployContract: React.FC<Props<T.DeployContract>>;
  FunctionCall: React.FC<Props<T.FunctionCall>>;
  Transfer: React.FC<Props<T.Transfer>>;
  Stake: React.FC<Props<T.Stake>>;
  AddKey: React.FC<Props<T.AddKey>>;
  DeleteKey: React.FC<Props<T.DeleteKey>>;
}

const COLLAPSE_ARGS_OPTIONS = {
  collapseText: "Show more",
  expandText: "Show less",
  minHeight: 200,
  maxHeight: 600
};

const transactionMessageRenderers: TransactionMessageRenderers = {
  CreateAccount: ({ transaction: { receiverId } }: Props<T.CreateAccount>) => (
    <>
      {"New account created: "}
      <AccountLink accountId={receiverId} />
    </>
  ),
  DeleteAccount: ({ transaction: { receiverId } }: Props<T.DeleteAccount>) => (
    <>
      {"Account deleted: "}
      <AccountLink accountId={receiverId} />
    </>
  ),
  DeployContract: ({
    transaction: { receiverId }
  }: Props<T.DeployContract>) => <>{`Contract deployed: ${receiverId}`}</>,
  FunctionCall: ({
    transaction: { receiverId },
    actionArgs,
    showDetails
  }: Props<T.FunctionCall>) => {
    let args;
    if (showDetails) {
      if (!actionArgs.args) {
        args = <p>Loading...</p>;
      } else {
        const decodedArgs = Buffer.from(actionArgs.args, "base64");
        let prettyArgs;
        try {
          const parsedJSONArgs = JSON.parse(decodedArgs.toString());
          prettyArgs = JSON.stringify(parsedJSONArgs, null, 2);
        } catch {
          prettyArgs = hexy(decodedArgs, { format: "twos" });
        }
        args = (
          <CodePreview collapseOptions={COLLAPSE_ARGS_OPTIONS}>
            {prettyArgs}
          </CodePreview>
        );
      }
    }

    return (
      <>
        {`Called method: '${actionArgs.method_name}' in contract: `}
        <AccountLink accountId={receiverId} />
        {showDetails ? (
          <dl>
            <dt>Arguments:</dt>
            <dd>{args}</dd>
          </dl>
        ) : null}
      </>
    );
  },
  Transfer: ({
    transaction: { receiverId },
    actionArgs: { deposit }
  }: Props<T.Transfer>) => (
    <>
      {`Transferred ${formatNEAR(deposit)} Ⓝ to `}
      <AccountLink accountId={receiverId} />
    </>
  ),
  Stake: ({ actionArgs: { stake, public_key } }: Props<T.Stake>) => (
    <>{`Staked: ${formatNEAR(stake)} Ⓝ ${public_key.substring(0, 15)}...`}</>
  ),
  AddKey: ({ transaction: { receiverId }, actionArgs }: Props<T.AddKey>) => (
    <>
      {typeof actionArgs.access_key.permission === "object" ? (
        <>
          {"Access key added for contract: "}
          <AccountLink
            accountId={
              actionArgs.access_key.permission.FunctionCall.receiver_id
            }
          />
        </>
      ) : (
        <>
          {"New key added for "}
          <AccountLink accountId={receiverId} />
          {`: ${actionArgs.public_key.substring(0, 15)}...`}
        </>
      )}
    </>
  ),
  DeleteKey: ({ actionArgs: { public_key } }: Props<T.DeleteKey>) => (
    <>{`Key deleted: ${public_key.substring(0, 15)}...`}</>
  )
};

export default (props: Props<AnyAction>) => {
  const MessageRenderer = transactionMessageRenderers[props.actionKind];
  if (MessageRenderer === undefined) {
    return (
      <>
        `${props.actionKind}: ${JSON.stringify(props.actionArgs)}`
      </>
    );
  }
  return <MessageRenderer {...props as any} showDetails={props.showDetails} />;
};
