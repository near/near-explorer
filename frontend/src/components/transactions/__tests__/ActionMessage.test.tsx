import renderer from "react-test-renderer";

import ActionMessage from "../ActionMessage";

const TRANSACTION = {
  hash: "BvJeW6gnFjkCBKCsRNEBrRLDQCFZNxLAi6uXzmLaVrrj",
  signerId: "signer.test",
  receiverId: "receiver.test",
  blockTimestamp: +new Date(2019, 1, 1)
};

describe("<ActionMessage />", () => {
  it("renders CreateAccount", () => {
    expect(
      renderer.create(
        <ActionMessage
          actionKind={"CreateAccount"}
          actionArgs={{}}
          transaction={TRANSACTION}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders DeleteAccount", () => {
    expect(
      renderer.create(
        <ActionMessage
          actionKind={"DeleteAccount"}
          actionArgs={{}}
          transaction={TRANSACTION}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders DeployContract", () => {
    expect(
      renderer.create(
        <ActionMessage
          actionKind={"DeployContract"}
          actionArgs={{}}
          transaction={TRANSACTION}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders FunctionCall", () => {
    expect(
      renderer.create(
        <ActionMessage
          actionKind={"FunctionCall"}
          actionArgs={{}}
          transaction={TRANSACTION}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders Transfer", () => {
    expect(
      renderer.create(
        <ActionMessage
          actionKind={"Transfer"}
          actionArgs={{ deposit: "1" }}
          transaction={TRANSACTION}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders Stake", () => {
    expect(
      renderer.create(
        <ActionMessage
          actionKind={"Stake"}
          actionArgs={{
            stake: "1",
            public_key: "ed25519:BgXFiJSzXz8VNFSW32rYNBiU7fUotKKeeDtPiSMkXMhx"
          }}
          transaction={TRANSACTION}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders AddKey", () => {
    expect(
      renderer.create(
        <ActionMessage
          actionKind={"AddKey"}
          actionArgs={{
            access_key: {},
            public_key: "ed25519:BgXFiJSzXz8VNFSW32rYNBiU7fUotKKeeDtPiSMkXMhx"
          }}
          transaction={TRANSACTION}
        />
      )
    ).toMatchSnapshot();
  });

  it("renders DeleteKey", () => {
    expect(
      renderer.create(
        <ActionMessage
          actionKind={"DeleteKey"}
          actionArgs={{
            public_key: "ed25519:BgXFiJSzXz8VNFSW32rYNBiU7fUotKKeeDtPiSMkXMhx"
          }}
          transaction={TRANSACTION}
        />
      )
    ).toMatchSnapshot();
  });
});
