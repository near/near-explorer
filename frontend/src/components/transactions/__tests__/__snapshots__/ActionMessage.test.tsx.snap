// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionMessage /> renders AddKey 1`] = `"New key added for @receiver.test: ed25519:BgXFiJS..."`;

exports[`<ActionMessage /> renders CreateAccount 1`] = `"New account created: @receiver.test"`;

exports[`<ActionMessage /> renders DeleteAccount 1`] = `"Account deleted: @receiver.test"`;

exports[`<ActionMessage /> renders DeleteKey 1`] = `"Key deleted: ed25519:BgXFiJS..."`;

exports[`<ActionMessage /> renders DeployContract 1`] = `"Contract deployed: receiver.test"`;

exports[`<ActionMessage /> renders FunctionCall 1`] = `"Called method in contract: \\"receiver.test\\""`;

exports[`<ActionMessage /> renders Stake 1`] = `"Staked: <0.00001 Ⓝ ed25519:BgXFiJS..."`;

exports[`<ActionMessage /> renders Transfer 1`] = `"Transferred <0.00001 Ⓝ to @receiver.test"`;
